File: 2022/21/solution.go
package main

import (
	"fmt"
	"math"
	"os"
	"path/filepath"
	"runtime"
	"strconv"
	"strings"
)

func main() {
	_, filename, _, _ := runtime.Caller(0)
	dirname := filepath.Dir(filename)
	inputFilePath := filepath.Join(dirname, "input.txt")
	data, err := os.ReadFile(inputFilePath)
	if err != nil {
		panic(err)
	}
	input := strings.Split(string(data), "\n")
	fmt.Println(part1(parseInput(input)))
	fmt.Println(part2(parseInput(input)))
}

type Monkey struct {
	value     float64
	monkey1   *Monkey
	monkey2   *Monkey
	operation string
}

func (m *Monkey) Value() float64 {
	if m.value > 0 { // assume no monkey has value zero
		return m.value
	}
	switch m.operation {
	case "+":
		return m.monkey1.Value() + m.monkey2.Value()
	case "*":
		return m.monkey1.Value() * m.monkey2.Value()
	case "-":
		return m.monkey1.Value() - m.monkey2.Value()
	case "/":
		return m.monkey1.Value() / m.monkey2.Value()
	}
	return -1
}

func parseInput(input []string) *map[string]*Monkey {
	monkeys := map[string]*Monkey{}
	for _, line := range input {
		parts := strings.Split(line, " ")
		curr := parts[0][:4]
		if _, ok := monkeys[curr]; !ok {
			monkey := Monkey{}
			monkeys[curr] = &monkey
		}
		if len(parts) > 2 {
			for _, name := range []string{parts[1], parts[3]} {
				if _, ok := monkeys[name]; !ok {
					monkey := Monkey{}
					monkeys[name] = &monkey
				}
			}
			monkeys[curr].monkey1 = monkeys[parts[1]]
			monkeys[curr].monkey2 = monkeys[parts[3]]
			monkeys[curr].operation = parts[2]
		} else {
			num, _ := strconv.Atoi(parts[1])
			monkeys[curr].value = float64(num)
		}
	}
	return &monkeys
}

func part1(monkeys *map[string]*Monkey) int {
	return int((*monkeys)["root"].Value())
}

func part2(monkeys *map[string]*Monkey) int {
	// find which side "humn" is on
	monkeyh := (*monkeys)["root"].monkey1
	monkey2 := (*monkeys)["root"].monkey2
	monkeyhvalue := monkeyh.Value()
	monkey2value := monkey2.Value()
	(*monkeys)["humn"].value = (*monkeys)["humn"].value * 2
	correlation := 1 // flag for positive/negative correlation
	monkeyhvalue2 := monkeyh.Value()
	if monkeyhvalue == monkeyhvalue2 {
		if monkey2.Value() < monkey2value {
			correlation = -1
		}
		monkeyh = monkey2
		monkey2 = (*monkeys)["root"].monkey1
	} else if monkeyhvalue2 < monkeyhvalue {
		correlation = -1
	}
	valueToMatch := monkey2.Value()
	lo := 1.0
	hi := math.MaxFloat64
	for lo < hi {
		mi := math.Floor(lo + (hi-lo)/2)
		(*monkeys)["humn"].value = mi
		monkeyhvalue = monkeyh.Value()
		if monkeyhvalue == valueToMatch {
			lo = mi
			break
		}
		if monkeyhvalue < valueToMatch {
			if correlation > 0 {
				lo = mi + 1
			} else {
				hi = mi - 1
			}
		} else {
			if correlation > 0 {
				hi = mi - 1
			} else {
				lo = mi + 1
			}
		}
	}
	return int(lo)
}

File: 2022/21/solution.js
const fs = require("fs");
const path = require("path");

const input = fs
  .readFileSync(path.join(__dirname, "input.txt"), "utf-8")
  .split("\n")
  .map((a) => a.split(": "));

const monkeys = new Map();

class Monkey {
  constructor(name, job) {
    this.name = name;
    this.type = "number";
    this.number;
    this.monkey1;
    this.monkey2;
    if (/\d/.test(job)) {
      this.number = Number(job);
    } else {
      const parts = job.split(" ");
      this.monkey1 = parts[0];
      this.type = parts[1];
      this.monkey2 = parts[2];
    }
  }

  yell() {
    switch (this.type) {
      case "number":
        return this.number;
      case "+":
        return (
          monkeys.get(this.monkey1).yell() + monkeys.get(this.monkey2).yell()
        );
      case "-":
        return (
          monkeys.get(this.monkey1).yell() - monkeys.get(this.monkey2).yell()
        );
      case "*":
        return (
          monkeys.get(this.monkey1).yell() * monkeys.get(this.monkey2).yell()
        );
      case "/":
        return (
          monkeys.get(this.monkey1).yell() / monkeys.get(this.monkey2).yell()
        );
    }
  }
}

for (const [name, job] of input) {
  monkeys.set(name, new Monkey(name, job));
}

console.log(monkeys.get("root").yell());

monkeys.get("root").type = "-";
let lo = 0;
let hi = 331120084396440 * 2;

while (lo < hi) {
  const mid = Math.floor(lo + (hi - lo) / 2);
  monkeys.get("humn").number = mid;
  const res = monkeys.get("root").yell();
  if (res === 0) {
    console.log(mid);
    break;
  }
  if (res > 0) {
    lo = mid + 1;
  } else {
    hi = mid - 1;
  }
}
